src/blackbox.py

```diff
@@ -0,0 +1,270 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Workload manager for Blackbox Exporter."""
+
+import logging
+import re
+import time
+import urllib.error
+import urllib.parse
+import urllib.request
+from typing import Optional
+
+import yaml
+from ops.framework import Object
+from ops.pebble import (  # type: ignore
+    ChangeError,
+    Layer,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class WorkloadManagerError(Exception):
+    """Base class for exceptions raised by WorkloadManager."""
+
+
+class ConfigUpdateFailure(WorkloadManagerError):
+    """Custom exception for failed config updates."""
+
+
+class ContainerNotReady(WorkloadManagerError):
+    """Raised when an operation is run that presumes the container being ready.."""
+
+
+class WorkloadManager(Object):
+    """Workload manager for blackbox exporter."""
+
+    _layer_name = _service_name = "blackbox-exporter"
+    _exe_name = "blackbox_exporter"
+
+    def __init__(
+        self,
+        charm,
+        *,
+        container_name: str,
+        port: int,
+        web_external_url: str,
+        config_path: str,
+    ):
+        # Must inherit from ops 'Object' to be able to register events.
+        super().__init__(charm, f"{self.__class__.__name__}-{container_name}")
+
+        self._unit = charm.unit
+
+        self._service_name = self._container_name = container_name
+        self._container = charm.unit.get_container(container_name)
+
+        self.api = BlackboxExporterApi(endpoint_url=charm._external_url)
+
+        self._port = port
+        self._web_external_url = web_external_url
+        self._config_path = config_path
+
+        # turn the container name to a valid Python identifier
+        snake_case_container_name = self._container_name.replace("-", "_")
+        charm.framework.observe(
+            charm.on[snake_case_container_name].pebble_ready, self._on_pebble_ready
+        )
+
+    @property
+    def is_ready(self):
+        """Is the workload ready to be interacted with?"""
+        return self._container.can_connect()
+
+    def _on_pebble_ready(self, _):
```

Comment:
We should also clear the workload version on stop.

Reply:
Not sure what you mean; this is how we set the workload version in our other charms as well

Reply:
We set version on pebble ready and update status, and clear it on stop.
We clear it on stop to be sure `juju status` isn't displaying incorrect information in case the upgrade sequence failed.

Reply:
Can you link an example?

Reply:
I couldn't find a reference of this but feel free to open an issue about it for a follow-up PR :)

Reply:
```
./mimir-worker-k8s-operator/src/charm.py:        self.unit.set_workload_version(self._mimir_version or "")
./mimir-k8s-operator/src/charm.py:        self.unit.set_workload_version("")
./karma-k8s-operator/src/charm.py:        self.unit.set_workload_version("")
./traefik-k8s-operator/src/charm.py:        self.unit.set_workload_version("")
```